File: sdk/go-holons/pkg/observability/observability.go
// Package observability is the Go reference implementation of the
// cross-SDK observability layer described in OBSERVABILITY.md. It
// provides structured logs, metrics, and lifecycle events with a
// zero-cost disabled mode.
//
// Activation follows the two-layer model from the spec:
//
//   - Layer 3 (env) — OP_OBS is a comma-separated list of families
//     (logs, metrics, events, prom) and is the sole source of on/off
//     per family. Parsed by FromEnv at process start.
//   - Layer 1 (code) — Configure supplies knobs (default log level,
//     Prometheus bind address, redacted field names) that apply
//     whenever a family is enabled.
//
// A single package-scope Observability holds the active state. Logger,
// Counter, Gauge, Histogram, and Events are all no-ops when their
// family is disabled.
package observability

import (
	"os"
	"strings"
	"sync"
	"time"
)

// Family identifies a group of signals the layer can emit.
type Family string

const (
	FamilyLogs    Family = "logs"
	FamilyMetrics Family = "metrics"
	FamilyEvents  Family = "events"
	FamilyProm    Family = "prom"
	FamilyOTel    Family = "otel" // reserved v2
)

// OP_OBS tokens that are valid in v1.
var v1Tokens = map[string]bool{
	"logs":    true,
	"metrics": true,
	"events":  true,
	"prom":    true,
	"all":     true,
}

// Config supplies knobs for when a family is enabled. None of these
// fields enables a family on their own — OP_OBS does.
type Config struct {
	// Slug of the holon (used as a well-known field on every signal).
	// Empty means the SDK will fall back to os.Args[0] basename.
	Slug string

	// Default log level applied to every new Logger. Per-logger
	// overrides use Logger.SetLevel. Zero value = INFO.
	DefaultLogLevel Level

	// Prometheus HTTP bind address when OP_OBS contains "prom".
	// ":0" means an ephemeral port. Ignored otherwise.
	PromAddr string

	// Names whose values are replaced by "<redacted>" in log fields
	// and event payloads before emission.
	RedactedFields []string

	// Ring buffer capacity for logs (default 1024) and events
	// (default 256).
	LogsRingSize   int
	EventsRingSize int

	// Run directory where stdout.log / events.jsonl / meta.json are
	// written. FromEnv derives this from OP_RUN_DIR + Slug +
	// InstanceUID because OP_RUN_DIR is the registry root.
	RunDir string

	// Instance identity from the supervisor. FromEnv fills these from
	// OP_INSTANCE_UID / OP_ORGANISM_UID / OP_ORGANISM_SLUG.
	InstanceUID  string
	OrganismUID  string
	OrganismSlug string
}

// Observability is the live state of the layer. The zero value is a
// no-op configuration with every family disabled — calls through it
// are safe and effectively free.
type Observability struct {
	cfg Config

	// Families that Layer 3 enabled. Immutable after Configure.
	families map[Family]bool

	// Lazy-built subsystems; nil when the corresponding family is off.
	registry *Registry
	bus      *EventBus
	ringLogs *LogRing
	loggers  sync.Map // map[string]*Logger

	// Redaction lookup.
	redact map[string]struct{}

	// Monotonic base for entry timestamps; locked once.
	startWall time.Time
}

// current is the package-scope instance returned by the constructors.
// Subsequent Configure / FromEnv calls replace it; the previous value
// is closed. A nil current means every accessor returns a no-op view.
var (
	currentMu sync.RWMutex
	current   *Observability
)

// Configure installs a new Observability with the given configuration
// and returns it. Families are read from OP_OBS at call time. Calling
// Configure again replaces the current instance; old loggers, metrics,
// and the event bus keep working against the previous instance until
// it is closed.
func Configure(cfg Config) *Observability {
	families := parseOPOBS(os.Getenv("OP_OBS"))

	// Defaults.
	if cfg.LogsRingSize <= 0 {
		cfg.LogsRingSize = 1024
	}
	if cfg.EventsRingSize <= 0 {
		cfg.EventsRingSize = 256
	}
	if cfg.DefaultLogLevel == LevelUnset {
		cfg.DefaultLogLevel = LevelInfo
	}
	if cfg.Slug == "" {
		cfg.Slug = defaultSlug()
	}

	obs := &Observability{
		cfg:       cfg,
		families:  families,
		redact:    toSet(cfg.RedactedFields),
		startWall: time.Now(),
	}

	if obs.families[FamilyLogs] {
		obs.ringLogs = NewLogRing(cfg.LogsRingSize)
	}
	if obs.families[FamilyMetrics] {
		obs.registry = NewRegistry()
	}
	if obs.families[FamilyEvents] {
		obs.bus = NewEventBus(cfg.EventsRingSize)
	}

	currentMu.Lock()
	current = obs
	currentMu.Unlock()
	return obs
}

// FromEnv is a convenience wrapper that applies Configure with fields
// populated from the environment: OP_INSTANCE_UID, OP_ORGANISM_UID,
// OP_ORGANISM_SLUG, OP_PROM_ADDR, OP_RUN_DIR. Any caller-provided
// fields in base override the env.
func FromEnv(base Config) *Observability {
	cfg := base
	if cfg.Slug == "" {
		cfg.Slug = defaultSlug()
	}
	if cfg.InstanceUID == "" {
		cfg.InstanceUID = os.Getenv("OP_INSTANCE_UID")
	}
	if cfg.OrganismUID == "" {
		cfg.OrganismUID = os.Getenv("OP_ORGANISM_UID")
	}
	if cfg.OrganismSlug == "" {
		cfg.OrganismSlug = os.Getenv("OP_ORGANISM_SLUG")
	}
	if cfg.PromAddr == "" {
		cfg.PromAddr = os.Getenv("OP_PROM_ADDR")
	}
	if cfg.RunDir == "" {
		if runRoot := os.Getenv("OP_RUN_DIR"); runRoot != "" && cfg.Slug != "" && cfg.InstanceUID != "" {
			if runDir, err := InstanceRunDir(runRoot, cfg.Slug, cfg.InstanceUID); err == nil {
				cfg.RunDir = runDir
			}
		}
	}
	return Configure(cfg)
}

// Current returns the active Observability. When none is configured,
// it returns a zero-value, disabled instance so callers don't need
// nil checks on the hot path.
func Current() *Observability {
	currentMu.RLock()
	defer currentMu.RUnlock()
	if current == nil {
		return disabled
	}
	return current
}

// Reset closes the current instance and installs a fresh disabled one.
// Intended for tests.
func Reset() {
	currentMu.Lock()
	current = nil
	currentMu.Unlock()
}

// disabled is a shared zero-value Observability with every family off.
// Pointers to it are returned by Current() when no instance is set.
var disabled = &Observability{families: map[Family]bool{}}

// Enabled reports whether the given family is active.
func (o *Observability) Enabled(f Family) bool {
	if o == nil {
		return false
	}
	return o.families[f]
}

// Slug returns the holon slug set on this instance (may be empty).
func (o *Observability) Slug() string {
	if o == nil {
		return ""
	}
	return o.cfg.Slug
}

// InstanceUID returns the instance UID set on this instance.
func (o *Observability) InstanceUID() string {
	if o == nil {
		return ""
	}
	return o.cfg.InstanceUID
}

// RunDir returns the per-instance registry directory, when configured.
func (o *Observability) RunDir() string {
	if o == nil {
		return ""
	}
	return o.cfg.RunDir
}

// OrganismUID returns the organism UID set on this instance.
func (o *Observability) OrganismUID() string { return o.cfg.OrganismUID }

// OrganismSlug returns the organism slug set on this instance.
func (o *Observability) OrganismSlug() string { return o.cfg.OrganismSlug }

// IsOrganismRoot reports whether this holon is the organism root.
// Canonical test from INSTANCES.md: OP_ORGANISM_UID == OP_INSTANCE_UID.
func (o *Observability) IsOrganismRoot() bool {
	if o == nil || o.cfg.OrganismUID == "" {
		return false
	}
	return o.cfg.OrganismUID == o.cfg.InstanceUID
}

// Close releases any resources held by the subsystems. Safe to call
// multiple times; subsequent calls are no-ops. Not reversible.
func (o *Observability) Close() error {
	if o == nil || o == disabled {
		return nil
	}
	if o.bus != nil {
		o.bus.Close()
	}
	// Logger ring buffer and registry are GC'd with the instance.
	return nil
}

// parseOPOBS turns the comma-separated env value into a family set.
// "all" expands to logs+metrics+events+prom. Unknown and v2-only
// tokens are dropped here so Configure can stay panic-free; CheckEnv
// applies the fail-fast startup policy.
func parseOPOBS(v string) map[Family]bool {
	out := map[Family]bool{}
	v = strings.TrimSpace(v)
	if v == "" {
		return out
	}
	for _, raw := range strings.Split(v, ",") {
		tok := strings.TrimSpace(raw)
		if tok == "" {
			continue
		}
		if isV2OnlyToken(tok) {
			continue
		}
		if !v1Tokens[tok] {
			// Unknown token. Same policy as above — the serve runner
			// surfaces this as a startup error via CheckEnv.
			continue
		}
		switch tok {
		case "all":
			out[FamilyLogs] = true
			out[FamilyMetrics] = true
			out[FamilyEvents] = true
			out[FamilyProm] = true
		default:
			out[Family(tok)] = true
		}
	}
	return out
}

// CheckEnv returns a non-nil error if OP_OBS contains an unknown or
// v2-only token, or if OP_SESSIONS is set in v1. Call once at process
// start to satisfy the fail-fast rule from OBSERVABILITY.md §Layer 3.
func CheckEnv() error {
	if v := strings.TrimSpace(os.Getenv("OP_SESSIONS")); v != "" {
		return &InvalidTokenError{
			Var:    "OP_SESSIONS",
			Token:  v,
			Reason: "sessions are reserved for v2; not implemented in v1",
		}
	}
	v := strings.TrimSpace(os.Getenv("OP_OBS"))
	if v == "" {
		return nil
	}
	for _, raw := range strings.Split(v, ",") {
		tok := strings.TrimSpace(raw)
		if tok == "" {
			continue
		}
		if tok == "otel" {
			return &InvalidTokenError{Token: tok, Reason: "otel export is reserved for v2; not implemented in v1"}
		}
		if tok == "sessions" {
			return &InvalidTokenError{Token: tok, Reason: "sessions are reserved for v2; not implemented in v1"}
		}
		if !v1Tokens[tok] {
			return &InvalidTokenError{Token: tok, Reason: "unknown OP_OBS token"}
		}
	}
	return nil
}

func isV2OnlyToken(tok string) bool {
	return tok == "otel" || tok == "sessions"
}

// InvalidTokenError is returned by CheckEnv when OP_OBS contains an
// invalid token.
type InvalidTokenError struct {
	Var    string
	Token  string
	Reason string
}

func (e *InvalidTokenError) Error() string {
	name := e.Var
	if name == "" {
		name = "OP_OBS"
	}
	return name + ": " + e.Reason + ": " + e.Token
}

func toSet(items []string) map[string]struct{} {
	if len(items) == 0 {
		return nil
	}
	out := make(map[string]struct{}, len(items))
	for _, s := range items {
		out[s] = struct{}{}
	}
	return out
}

// defaultSlug derives a best-effort slug from os.Args[0]'s basename.
// Used when the caller did not set Config.Slug explicitly.
func defaultSlug() string {
	if len(os.Args) == 0 {
		return ""
	}
	base := os.Args[0]
	if i := strings.LastIndexByte(base, '/'); i >= 0 {
		base = base[i+1:]
	}
	return base
}

File: sdk/js-web-holons/src/observability.mjs
/**
 * Browser / WASM reference implementation of the cross-SDK observability layer.
 *
 * Smaller scope than Node (js-holons): no filesystem access,
 * no Prometheus HTTP server — js-web holons are dial-only sandboxed
 * peers per the spec §Transports without outbound dial. The root
 * (Hub or Flutter organism) reads their HolonObservability.Logs /
 * Events / Metrics over the existing WebSocket full-duplex channel.
 *
 * Same activation model (OP_OBS env through import.meta.env or a
 * runtime bootstrap), same public surface, same on-wire shape.
 */

export const Family = Object.freeze({
  LOGS: 'logs', METRICS: 'metrics', EVENTS: 'events', PROM: 'prom', OTEL: 'otel',
});

const V1_TOKENS = new Set(['logs', 'metrics', 'events', 'prom', 'all']);

export class InvalidTokenError extends Error {
  constructor(token, reason) {
    super(`OP_OBS: ${reason}: ${token}`);
    this.name = 'InvalidTokenError';
    this.token = token;
  }
}

export function parseOpObs(raw) {
  const out = new Set();
  if (!raw || !raw.trim()) return out;
  for (const p of raw.split(',')) {
    const tok = p.trim();
    if (!tok) continue;
    if (tok === 'otel' || tok === 'sessions') continue;
    if (!V1_TOKENS.has(tok)) continue;
    if (tok === 'all') {
      out.add(Family.LOGS); out.add(Family.METRICS);
      out.add(Family.EVENTS); out.add(Family.PROM);
    } else {
      out.add(tok);
    }
  }
  return out;
}

export function checkEnv(env = {}) {
  if ((env.OP_SESSIONS || '').trim()) {
    throw new InvalidTokenError(env.OP_SESSIONS.trim(), 'sessions are reserved for v2; not implemented in v1');
  }
  const raw = (env.OP_OBS || '').trim();
  if (!raw) return;
  for (const p of raw.split(',')) {
    const tok = p.trim();
    if (!tok) continue;
    if (tok === 'otel') throw new InvalidTokenError(tok, 'otel export is reserved for v2; not implemented in v1');
    if (tok === 'sessions') throw new InvalidTokenError(tok, 'sessions are reserved for v2; not implemented in v1');
    if (!V1_TOKENS.has(tok)) throw new InvalidTokenError(tok, 'unknown OP_OBS token');
  }
}

export const Level = Object.freeze({
  UNSET: 0, TRACE: 1, DEBUG: 2, INFO: 3, WARN: 4, ERROR: 5, FATAL: 6,
});
const LEVEL_NAMES = { 1: 'TRACE', 2: 'DEBUG', 3: 'INFO', 4: 'WARN', 5: 'ERROR', 6: 'FATAL' };

export function parseLevel(s) {
  if (!s) return Level.INFO;
  const u = String(s).trim().toUpperCase();
  return ({ TRACE: Level.TRACE, DEBUG: Level.DEBUG, INFO: Level.INFO,
    WARN: Level.WARN, WARNING: Level.WARN, ERROR: Level.ERROR,
    FATAL: Level.FATAL })[u] || Level.INFO;
}

export const EventType = Object.freeze({
  UNSPECIFIED: 0,
  INSTANCE_SPAWNED: 1, INSTANCE_READY: 2, INSTANCE_EXITED: 3, INSTANCE_CRASHED: 4,
  SESSION_STARTED: 5, SESSION_ENDED: 6, HANDLER_PANIC: 7, CONFIG_RELOADED: 8,
});

export function appendDirectChild(src, childSlug, childUid) {
  return [...(src || []), { slug: childSlug, instance_uid: childUid }];
}
export function enrichForMultilog(wire, srcSlug, srcUid) {
  return appendDirectChild(wire, srcSlug, srcUid);
}

export class LogRing {
  constructor(capacity = 1024) {
    this._capacity = Math.max(1, capacity);
    this._buf = [];
    this._subs = [];
  }
  push(e) {
    this._buf.push(e);
    if (this._buf.length > this._capacity) this._buf.splice(0, this._buf.length - this._capacity);
    for (const fn of [...this._subs]) try { fn(e); } catch (_) {}
  }
  drain() { return [...this._buf]; }
  drainSince(cutoff) { return this._buf.filter((e) => e.timestamp >= cutoff); }
  subscribe(fn) {
    this._subs.push(fn);
    return () => { const i = this._subs.indexOf(fn); if (i >= 0) this._subs.splice(i, 1); };
  }
}

export class EventBus {
  constructor(capacity = 256) {
    this._capacity = Math.max(1, capacity);
    this._buf = [];
    this._subs = [];
    this._closed = false;
  }
  emit(e) {
    if (this._closed) return;
    this._buf.push(e);
    if (this._buf.length > this._capacity) this._buf.splice(0, this._buf.length - this._capacity);
    for (const fn of [...this._subs]) try { fn(e); } catch (_) {}
  }
  drain() { return [...this._buf]; }
  drainSince(cutoff) { return this._buf.filter((e) => e.timestamp >= cutoff); }
  subscribe(fn) {
    this._subs.push(fn);
    return () => { const i = this._subs.indexOf(fn); if (i >= 0) this._subs.splice(i, 1); };
  }
  close() { this._closed = true; this._subs.length = 0; }
}

export class Counter {
  constructor(name, help = '', labels = {}) {
    this.name = name; this.help = help; this.labels = { ...labels };
    this._v = 0;
  }
  inc(n = 1) { if (n >= 0) this._v += n; }
  add(n) { this.inc(n); }
  value() { return this._v; }
}

export class Gauge {
  constructor(name, help = '', labels = {}) {
    this.name = name; this.help = help; this.labels = { ...labels };
    this._v = 0;
  }
  set(v) { this._v = Number(v); }
  add(d) { this._v += Number(d); }
  value() { return this._v; }
}

export const DEFAULT_BUCKETS = [
  50e-6, 100e-6, 250e-6, 500e-6,
  1e-3, 2.5e-3, 5e-3, 10e-3, 25e-3, 50e-3, 100e-3, 250e-3, 500e-3,
  1.0, 2.5, 5.0, 10.0, 30.0, 60.0,
];

export class Histogram {
  constructor(name, help = '', labels = {}, bounds = null) {
    this.name = name; this.help = help; this.labels = { ...labels };
    this._bounds = [...(bounds && bounds.length ? bounds : DEFAULT_BUCKETS)].sort((a, b) => a - b);
    this._counts = new Array(this._bounds.length).fill(0);
    this._total = 0;
    this._sum = 0;
  }
  observe(v) {
    this._total += 1;
    this._sum += v;
    for (let i = 0; i < this._bounds.length; i++) if (v <= this._bounds[i]) this._counts[i] += 1;
  }
  snapshot() { return { bounds: [...this._bounds], counts: [...this._counts], total: this._total, sum: this._sum }; }
  static quantile(snap, q) {
    if (snap.total === 0) return NaN;
    const target = snap.total * q;
    for (let i = 0; i < snap.counts.length; i++) if (snap.counts[i] >= target) return snap.bounds[i];
    return Infinity;
  }
}

function metricKey(name, labels) {
  const keys = Object.keys(labels || {}).sort();
  return keys.length ? name + '|' + keys.map((k) => `${k}=${labels[k]}`).join(',') : name;
}

export class Registry {
  constructor() {
    this._c = new Map(); this._g = new Map(); this._h = new Map();
  }
  counter(name, help = '', labels = {}) {
    const k = metricKey(name, labels);
    let c = this._c.get(k);
    if (!c) { c = new Counter(name, help, labels); this._c.set(k, c); }
    return c;
  }
  gauge(name, help = '', labels = {}) {
    const k = metricKey(name, labels);
    let g = this._g.get(k);
    if (!g) { g = new Gauge(name, help, labels); this._g.set(k, g); }
    return g;
  }
  histogram(name, help = '', labels = {}, bounds = null) {
    const k = metricKey(name, labels);
    let h = this._h.get(k);
    if (!h) { h = new Histogram(name, help, labels, bounds); this._h.set(k, h); }
    return h;
  }
}

export class Logger {
  constructor(obs, name) {
    this._obs = obs;
    this._name = name;
    this._level = obs ? obs.cfg.defaultLogLevel : Level.FATAL;
  }
  setLevel(l) { this._level = l; }
  enabled(l) { return this._obs && l >= this._level; }
  _log(level, message, fields) {
    if (!this.enabled(level)) return;
    const redact = this._obs.cfg.redactedFields || new Set();
    const out = {};
    if (fields) {
      for (const k of Object.keys(fields)) {
        const has = redact.has ? redact.has(k) : (redact.includes && redact.includes(k));
        out[k] = has ? '<redacted>' : String(fields[k] ?? '');
      }
    }
    const entry = {
      timestamp: Date.now() / 1000,
      level, slug: this._obs.cfg.slug, instance_uid: this._obs.cfg.instanceUid,
      session_id: '', rpc_method: '', message, fields: out, caller: '', chain: [],
    };
    if (this._obs.logRing) this._obs.logRing.push(entry);
  }
  trace(m, f) { this._log(Level.TRACE, m, f); }
  debug(m, f) { this._log(Level.DEBUG, m, f); }
  info(m, f)  { this._log(Level.INFO,  m, f); }
  warn(m, f)  { this._log(Level.WARN,  m, f); }
  error(m, f) { this._log(Level.ERROR, m, f); }
  fatal(m, f) { this._log(Level.FATAL, m, f); }
}

const _disabledLogger = Object.freeze({
  _obs: null, _name: '', _level: Level.FATAL,
  enabled() { return false; }, setLevel() {},
  trace() {}, debug() {}, info() {}, warn() {}, error() {}, fatal() {},
});

export class Observability {
  constructor(cfg, families) {
    this.cfg = cfg;
    this.families = families;
    this.logRing = families.has(Family.LOGS) ? new LogRing(cfg.logsRingSize || 1024) : null;
    this.registry = families.has(Family.METRICS) ? new Registry() : null;
    this.eventBus = families.has(Family.EVENTS) ? new EventBus(cfg.eventsRingSize || 256) : null;
    this._loggers = new Map();
  }
  enabled(f) { return this.families.has(f); }
  isOrganismRoot() { return !!this.cfg.organismUid && this.cfg.organismUid === this.cfg.instanceUid; }
  logger(name) {
    if (!this.families.has(Family.LOGS)) return _disabledLogger;
    let l = this._loggers.get(name);
    if (!l) { l = new Logger(this, name); this._loggers.set(name, l); }
    return l;
  }
  counter(n, h = '', labels = {}) { return this.registry ? this.registry.counter(n, h, labels) : null; }
  gauge(n, h = '', labels = {}) { return this.registry ? this.registry.gauge(n, h, labels) : null; }
  histogram(n, h = '', labels = {}, bounds = null) { return this.registry ? this.registry.histogram(n, h, labels, bounds) : null; }
  emit(type, payload) {
    if (!this.eventBus) return;
    const redact = this.cfg.redactedFields || new Set();
    const p = {};
    if (payload) {
      for (const k of Object.keys(payload)) {
        const has = redact.has ? redact.has(k) : (redact.includes && redact.includes(k));
        p[k] = has ? '<redacted>' : String(payload[k] ?? '');
      }
    }
    this.eventBus.emit({
      timestamp: Date.now() / 1000,
      type, slug: this.cfg.slug, instance_uid: this.cfg.instanceUid,
      session_id: '', payload: p, chain: [],
    });
  }
  close() { if (this.eventBus) this.eventBus.close(); }
}

let _current = null;

export function configure(cfg = {}) {
  const env = (typeof import.meta !== 'undefined' && import.meta.env) || (typeof globalThis !== 'undefined' && globalThis.__HOLON_ENV__) || {};
  const families = parseOpObs(env.OP_OBS || '');
  const normalized = {
    slug: cfg.slug || '',
    instanceUid: cfg.instanceUid || env.OP_INSTANCE_UID || '',
    organismUid: cfg.organismUid || env.OP_ORGANISM_UID || '',
    organismSlug: cfg.organismSlug || env.OP_ORGANISM_SLUG || '',
    promAddr: cfg.promAddr || env.OP_PROM_ADDR || '',
    runDir: cfg.runDir || env.OP_RUN_DIR || '',
    defaultLogLevel: cfg.defaultLogLevel || Level.INFO,
    redactedFields: cfg.redactedFields ? new Set(cfg.redactedFields) : new Set(),
    logsRingSize: cfg.logsRingSize,
    eventsRingSize: cfg.eventsRingSize,
  };
  const obs = new Observability(normalized, families);
  _current = obs;
  return obs;
}

export function fromEnv(base = {}) { return configure(base); }

export function current() {
  if (_current) return _current;
  return new Observability({
    slug: '', instanceUid: '', organismUid: '', organismSlug: '',
    defaultLogLevel: Level.FATAL, redactedFields: new Set(),
  }, new Set());
}

export function reset() { if (_current) _current.close(); _current = null; }
